### PR TITLE
WT-11062 Safe free the ref addr to allow concurrent access

### PR DESCRIPTION
**Pull request by kommiharibabu**
When the ref is being accessed concurrently, it will be saved until the access is finished to free it safely.

### CODE REVIEW

**Comment by kommiharibabu** [comment]
@agorrod and @ajmorton This approach doesn't free the `ref->addr` if anyone is accessing it concurrently, but it sets it to NULL. This approach addresses the problem with this ticket. Later, when no one is using this address, it will be freed. This memory is tracked with `split_gen` generation number.

File: src/btree/bt_discard.c [hunk 1]
```diff
@@ -280,10 +282,15 @@ __wt_ref_addr_free(WT_SESSION_IMPL *session, WT_REF *ref)
     }
 
     if (home == NULL || __wt_off_page(home, ref_addr)) {
-        ((WT_ADDR *)ref_addr)->size = 0;
-        __wt_free(session, ((WT_ADDR *)ref_addr)->addr);
-        __wt_free(session, ref_addr);
+        split_gen = __wt_gen(session, WT_GEN_SPLIT);
+        WT_ASSERT_ALWAYS(session, split_gen != 0, "Must be inside the split generation.");
+        WT_TRET(__wt_split_safe_free(
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/btree/bt_discard.c]: Thanks. I made the change as suggested instead of making a split function a public function.

File: src/include/btree_inline.h [hunk 2]
```diff
@@ -2344,6 +2336,7 @@ __wt_btcur_skip_page(
      * if the page has been modified since it was reconciled, since neither the delete information
      * nor the timestamp information is necessarily up to date.
      */
+    WT_ASSERT(session, __wt_gen(session, WT_GEN_SPLIT) != 0);
```
Comment (agorrod) [inline src/include/btree_inline.h:2339]: Could this be moved to a WT_ASSERT_ALWAYS within the __wt_ref_addr_copy implementation? It might need to go after the check that the addr is non-nul.

Comment (kommiharibabu) [reply src/include/btree_inline.h]: The `__wt_ref_addr_copy` is called in locations where the split generation is not taken, so moving the assert can fail. This is the only location where we can get into problems, so placing the assert here would be good.

File: src/btree/bt_discard.c [hunk 3]
```diff
@@ -240,6 +240,28 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
...
+ *     otherwise.
+ */
+static void
+__ref_addr_safe_free(WT_SESSION_IMPL *session, void *ref_addr)
+{
+    WT_DECL_RET;
+    uint64_t split_gen;
+
+    split_gen = __wt_gen(session, WT_GEN_SPLIT);
+    WT_ASSERT_ALWAYS(session, split_gen != 0, "Must be inside the split generation.");
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/btree/bt_discard.c]: Thanks, changed.

File: src/btree/bt_discard.c [hunk 4]
```diff
@@ -240,6 +240,28 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
     __wt_free(session, page->modify);
 }
 
+/*
+ * __ref_addr_safe_free --
+ *     Free the ref address if we can be sure no thread is accessing it, or schedule it to be freed
+ *     otherwise.
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/btree/bt_discard.c]: Thanks, changed as suggested.

File: src/btree/bt_discard.c [hunk 5]
```diff
@@ -240,6 +240,28 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
...
+ */
+static void
+__ref_addr_safe_free(WT_SESSION_IMPL *session, void *ref_addr)
+{
+    WT_DECL_RET;
+    uint64_t split_gen;
+
+    split_gen = __wt_gen(session, WT_GEN_SPLIT);
```
[the remark this answers is not in the document]
Comment (agorrod) [reply src/btree/bt_discard.c:255]: Ah, I see. Thanks.

Do you think this also needs the WT_FULL_BARRIER? If it does, we might need two different version of this function: one that is called in the tight loop freeing a bunch of refs (which only need a single split generation), and one which is a one-off free that doesn't want to fuss with doing the barrier/generation dance.

[the remark this answers is not in the document]
Comment (agorrod) [reply src/btree/bt_discard.c:255]: I'm also happy to be convinced that a barrier isn't necessary, I didn't think that through to the ground.

[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/btree/bt_discard.c:255]: In the split logic, we save the `split_gen` in the parent page. Due to this reason, we do not want the code to be reordered, which means replacing the old page index with a new one before the `split_gen` save operation. In this ref address-free scenario, we already have the logic of replacing the `ref->addr` with NULL using an atomic operation. I do not think that we need a full barrier here.

[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/btree/bt_discard.c:255]: To avoid incrementing the `split_gen` for every ref address free, how about moving the `gen_next` into the caller and incrementing it only once? Also, it is worth mentioning that the address is freed only when it is `off page`.

[the remark this answers is not in the document]
Comment (agorrod) [reply src/btree/bt_discard.c:255]: Ack. The call to __wt_gen is pretty lightweight as is It just retrieves a shared variable wihout any coordination.

I think the code is fine as-is since if it doesn't require additional coordination.

**Review by agorrod — APPROVED** [review]
LGTM - though it would be great if you can bring some other people along on understanding this change.

Also: Please run stress testing and ASAN testing before merging.

File: src/btree/bt_discard.c [hunk 6]
```diff
@@ -240,6 +240,28 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
...
+    WT_TRET(__wt_stash_add(
+      session, WT_GEN_SPLIT, split_gen, ((WT_ADDR *)ref_addr)->addr, ((WT_ADDR *)ref_addr)->size));
+    WT_TRET(__wt_stash_add(session, WT_GEN_SPLIT, split_gen, ref_addr, sizeof(WT_ADDR)));
+    __wt_gen_next(session, WT_GEN_SPLIT, NULL);
```
Comment (ajmorton) [inline src/btree/bt_discard.c:259]: Can we have a comment here (or on the `wt_gen` call) explaining that no splits are taking place, but we're using the generation since we know readers must hold one?

File: src/include/btree_inline.h [hunk 7]
```diff
@@ -2352,6 +2344,8 @@ __wt_btcur_skip_page(
      * if the page has been modified since it was reconciled, since neither the delete information
      * nor the timestamp information is necessarily up to date.
      */
```
Comment (ajmorton) [inline src/include/btree_inline.h:2346]: Similarly here it'd be useful to know that `__ref_addr_safe_free` uses the generation when freeing the ref, hence the need for a readers to be in a generation

**Review by ajmorton — COMMENTED** [review]
TL:DR I'm happy with this change but I've written small novel to convince myself of the fact. @kommiharibabu  I'll request a few more comments in the PR to act as breadcrumbs for the next time we read this code. 
CC @luke-pearson  for a second opinion, in particular on the reasoning that `r2 happens before f1` is correct.

<br/>
<br/>  

First up a key point: While we're using `WT_GEN_SPLIT` there are no page splits taking place. The reading thread is always inside a split generation when it reads the ref, so we make use of that fact rather than create a whole new generation counter for the `__ref_addr_safe_free` behaviour.

**Scenario**
We have one thread freeing the ref and another thread reading the ref,  and we need to make sure the freeing thread never destroys the ref out from under the reader thread. Heavily simplified pseudocode for the new behaviour of the two threads looks like
```c
// Freeing thread
cas(ref = NULL)        // line f1
f_gen = wt_gen()       // line f2
stash(ref, f_gen)      // line f3

// Reading thread
r_gen = wt_gen_enter() // line r1
r_ref = ref            // line r2
r_addr = r_ref->addr   // line r3
```
For the freeing thread
- All of our compare_and_swaps are done with `__ATOMIC_SEQ_CST`. This guarantees all later instructions cannot be reordered before the CAS.
- `wt_gen` is a non-atomic read from `conn->generation`. In this case it's not an issue (the CAS gives us the required ordering) but in future it'd be nice to make this an atomic read (WT-11907 has some further context on this)
- When we add something to the stash we also provide a generation. The resource will only be freed after all sessions have left that generation and are finished with the resource.

For the reading thread:
- `wt_gen_enter` reads the current generation [in a while loop using a full barrier](https://github.com/wiredtiger/wiredtiger/blob/8f302aff08a74d783883d117ef8b15b2035abbac/src/support/generation.c#L373). This guarantees the entered `session->generation <= conn->generation`, and that reading `r_ref` takes place after the generation has been entered. We also have an assert in the reader code to ensure this.

<br/>  
<br/>  

The fail state we need to protect against is the following
1. The reader threads reads the ref as non-`NULL` (`line r2`)
2. The freeing thread nulls the ref and adds it to the stash (`lines f1-f3`). The stashed ref is then freed
3. The reader thread attempts to access the `NULL`ed ref (`line r3`)

We can guarantee the above never happens if we can prove that `r_gen` (`line r1`) is always less than or equal to `f_gen` (`line f2`), as this prevents the ref from being freed until the reader is finished using it.

The ordering constraints we have are:
1. In the reader thread `r_gen` must be read in **before** `r_ref`  due to the full barrier in `wt_gen_enter`. 
   **r1 happens before r2**
2. For our fail case the reader thread must see the state of `ref` **before** the freeing thread nulls it. ref isn't a traditional synchronisation variable with acquire/release semantics, but since both of our threads are using full barriers I think this is acceptable. 
   **r2 happens before f1**
3. In the freeing thread `f_gen` must be read in **after** the ref has been nulled due to the full barrier from our `__ATOMIC_SEQ_CST` compare_and_swap. 
   **f1 happens before f2**

This gives a transitive **r1 happens before f2**:
```c
r_gen = wt_gen_enter() // line r1
f_gen = wt_gen()       // line f2
```
and we know that `__wt_gen_enter` guarantees the `session->generation` (`r_gen`) is less than or equal to the `connection->generation` subsequently read into `f_gen`. Therefore `r_gen <= f_gen` satisfying our constraint and making the change safe.

File: src/btree/bt_discard.c [hunk 8]
```diff
@@ -240,6 +240,28 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
     __wt_free(session, page->modify);
 }
 
+/*
+ * __ref_addr_safe_free --
+ *     Any thread that is reviewing the address in a WT_REF, must also be holding a split generation
+ *     to ensure that the page index they are using remains valid. Use that same split generation to
```
Comment (luke-pearson) [inline src/btree/bt_discard.c:246]: This does seem contradictory to what is in the code, similar to what Alex G noted below. If we were using the same generation we should call "__wt_session_gen" should we improve the comment or am I misunderstanding something?

File: src/include/btree_inline.h [hunk 9]
```diff
@@ -2352,6 +2344,8 @@ __wt_btcur_skip_page(
...
+    WT_ASSERT_ALWAYS(session, __wt_session_gen(session, WT_GEN_SPLIT) != 0,
```
Comment (luke-pearson) [inline src/include/btree_inline.h:2347]: Shouldn't we build this assertion into `__wt_ref_addr_copy`? Also the comment on this assertion confuses me as it is copying an addr not freeing it.

Comment (kommiharibabu) [reply src/include/btree_inline.h:2347]: No. The `__wt_ref_addr_copy` is called from locations where the split generation is not acquired. Due to this reason, we cannot move this assert into that function. Otherwise, all the callers have to take the split generation when calling this function.

File: src/btree/bt_discard.c (hunk 8 above)
Comment (kommiharibabu) [reply src/btree/bt_discard.c:246]: The same split generation means `WT_GEN_SPLIT` using here and not the same generation number. The use of `__wt_gen` is correct. I will check if I can improve the comment to avoid this confusion.

File: src/include/btree_inline.h (hunk 9 above)
Comment (luke-pearson) [reply src/include/btree_inline.h:2347]: I think we should require all callers to take the relevant generation, if that requires creating a WT_REF_ADDR generation then I'd be okay with that. Otherwise this is a half-baked usage of generations where some functions require that semantic and others don't but there's no obvious distinction between the tow.

File: src/btree/bt_discard.c (hunk 8 above)
Comment (luke-pearson) [reply src/btree/bt_discard.c:246]: Perhaps: "Utilize the same generation type to safely free the address once all users of it have left the generation."

Comment (kommiharibabu) [reply src/btree/bt_discard.c]: Thanks. Fixed as suggested.

File: src/btree/bt_discard.c (hunk 6 above)
Comment (kommiharibabu) [reply src/btree/bt_discard.c:264]: I tried to add some comments to explain why WT_GEN_SPLIT is used, please check and let me know.

File: src/include/btree_inline.h (hunk 7 above)
Comment (kommiharibabu) [reply src/include/btree_inline.h]: The function `__ref_addr_safe_free` header has the comments explaining the same, let me know If you want any more details to be added.

File: src/include/btree_inline.h (hunk 9 above)
Comment (kommiharibabu) [reply src/include/btree_inline.h]: Thanks. I find there are not many places where we need to add the split generation. So I went ahead with WT_GEN_SPLIT in the callers.

**Comment by luke-pearson** [comment]
Elaborating on what @ajmorton was discussing I think this is correct.

```
// Freeing thread
cas(ref->addr = NULL)               // line f1
split_gen = conn->split_gen         // line f2
stash(ref->addr, split_gen)         // line f3

// Reading thread
session_split_gen = wt_gen_enter()  // line r1
r_addr = r_ref->addr                // line r2
```
In this example the bad state could occur for two reasons:
1.  If line `r2` happened before `r1` allowing the reader thread to read the value of ref->addr prior to truly acquiring a generation. Thankfully we know generation acquisition has a `FULL_BARRIER` so `r2` must happen after that.
2. As Andrew also said the generation read on `f2` must happen after `f1`, per the CAS logic. So the `conn->split_gen` read cannot happen earlier preventing it from reading an older `conn->split_gen`.

File: src/include/btree_inline.h (hunk 9 above)
Comment (luke-pearson) [reply src/include/btree_inline.h]: Thanks Hari I think that makes this a lot easier to reason about!

File: src/btree/bt_prefetch.c [hunk 10]
```diff
@@ -89,13 +89,15 @@ __wt_prefetch_page_in(WT_SESSION_IMPL *session, WT_PREFETCH_QUEUE_ENTRY *pe)
 
     WT_STAT_CONN_INCR(session, block_prefetch_pages_read);
 
+    WT_ENTER_GENERATION(session, WT_GEN_SPLIT);
     if (__wt_ref_addr_copy(session, pe->ref, &addr)) {
         WT_ERR(__wt_page_in(session, pe->ref, WT_READ_PREFETCH));
         WT_ERR(__wt_page_release(session, pe->ref, 0));
     } else
-        return (WT_ERROR);
+        ret = (WT_ERROR);
```
Comment (luke-pearson) [inline src/btree/bt_prefetch.c:97]: I wonder if the original code here had a bug? This changes the behaviour slightly as now we'll call `__wt_page_release` and return `WT_ERROR` whereas previously we would just return `WT_ERROR`. What do you think Hari?

File: src/btree/bt_discard.c (hunk 6 above)
Comment (ajmorton) [reply src/btree/bt_discard.c:264]: Looks good, thanks!

File: src/btree/bt_prefetch.c (hunk 10 above)
Comment (luke-pearson) [reply src/btree/bt_prefetch.c:97]: @mm-ng

Comment (mm-ng) [reply src/btree/bt_prefetch.c:97]: Ok, so the code you have here is outdated. We've changed it back to `return (WT_ERROR)` rather than `ret = (WT_ERROR)`.

Comment (kommiharibabu) [reply src/btree/bt_prefetch.c:107]: I merged the latest develop code that does not have another page release function call.

**Review by ajmorton — APPROVED** [review]
LGTM pending the discussion about ` ret = (WT_ERROR)` in `__wt_prefetch_page_in` 👍

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Comment by kommiharibabu** [comment]
evergreen merge
